### PR TITLE
EVP & TLS: Add necessary EC_KEY data extraction functions, and use them

### PR DESCRIPTION
**Pull request by levitte**
libssl code uses EVP_PKEY_get0_EC_KEY() to extract certain basic data
from the EC_KEY.  We replace that with EVP_PKEY functions.

This may or may not be refactored later on.

### CODE REVIEW

**Comment by levitte** [comment]
If nothing else, this can be an interim solution until #11340 has been merged.
#11328 depends on something like this.

**Comment by mattcaswell** [comment]
Hmmm. I can't say I'm keen on this. This seems to just sidestep the problem of what to do with the `EVP_PKEY_get0_EC_KEY` et al functions - and fixes it *for us*. But does so by polluting the EVP_PKEY API with algorithm specific functions.

**Comment by levitte** [comment]
Yes, that's true.  However, as soon as we get EC key generation, we get issues with this, so the question is how long that should be stuck waiting for "the right thing" to happen.

I *could* make those functions internal but still exported, if that helps matters

**Comment by levitte** [comment]
This *is* an interim solution

**Comment by levitte** [comment]
Marked as WIP for obvious reasons

**Comment by mattcaswell** [comment]
> I could make those functions internal but still exported, if that helps matters

That would definitely help I think.

**Comment by levitte** [comment]
Although, I think `EVP_PKEY_can_sign()` is more generally useful...

**Comment by levitte** [comment]
There.  I've also added EVP_PKEY_is_a(), which has been floating a more than one branch I juggled with

File: doc/man3/EVP_PKEY_is_a.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_is_a.pod]: I think some examples of valid names would be useful.

File: ssl/statem/statem_lib.c
Comment (mattcaswell) [inline ssl/statem/statem_lib.c]: Err...fprintf?

File: ssl/statem/statem_lib.c
Comment (mattcaswell) [inline ssl/statem/statem_lib.c]: Shouldn't this be `evp_pkey_get_EC_KEY_curve_nid` (i.e. in lowercase)?

File: ssl/t1_lib.c
Comment (mattcaswell) [inline ssl/t1_lib.c]: evp_pkey_...and so on for other occurrences

File: include/internal/evp.h
Comment (mattcaswell) [inline include/internal/evp.h]: Should there be some TODO comment somewhere saying this is an interim solution?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: include/internal/evp.h
Comment (mattcaswell) [inline include/internal/evp.h]: Should be `OSSL_INTERNAL_EVP_H`

File: ssl/statem/statem_lib.c
Comment (levitte) [reply ssl/statem/statem_lib.c]: Oops, debug stuff.  Will remove

File: ssl/statem/statem_lib.c
Comment (levitte) [reply ssl/statem/statem_lib.c]: Yes.  There are a few of these, I've already corrected them locally

File: include/internal/evp.h
Comment (levitte) [reply include/internal/evp.h]: Yup, already fixed locally

File: include/internal/evp.h
Comment (levitte) [reply include/internal/evp.h]: I thought I already added such a comment???  probably somewhere else, and yeah I can duplicate it here

File: doc/man3/EVP_PKEY_is_a.pod
Comment (levitte) [reply doc/man3/EVP_PKEY_is_a.pod]: Sure

**Comment by mattcaswell** [comment]
Could you push the latest version of your local branch here? I'm experimenting with your ec key gen stuff...

**Comment by levitte** [comment]
Done, and in #11328 as well

**Comment by levitte** [comment]
Rebased on master, now that #11375 is in.

File: ssl/statem/statem_lib.c
Comment (levitte) [reply ssl/statem/statem_lib.c]: Removed them all

File: include/internal/evp.h
Comment (levitte) [reply include/internal/evp.h]: Done

File: doc/man3/EVP_PKEY_is_a.pod
Comment (levitte) [reply doc/man3/EVP_PKEY_is_a.pod]: Done

**Comment by levitte** [comment]
Ping!

**Comment by levitte** [comment]
Travis failure re documentation:

```
doc/man3/EVP_PKEY_is_a.pod:1: EXAMPLE not EXAMPLES section.
```

That document contains *two* examples, so I'm starting to wonder if find-doc-nits isn't a little *too* picky in this case.

**Comment by levitte** [comment]
Ping!

**Comment by mattcaswell** [comment]
Needs rebase!

**Comment by t8m** [comment]
There are now conflicts and the Travis failure is probably relevant

**Comment by levitte** [comment]
Just rebased

File: ssl/t1_lib.c
Comment (mattcaswell) [inline ssl/t1_lib.c]: This function is inextricably linked with the point encoding/decoding. When we push the point encoding/decoding provider side then we will also have to push the logic about what point compression is supported provider side too, i.e. we should ask the provider what point compression is supported for any particular key - rather than trying to infer if from the field type. That works for own stuff because we know we support point compression - but we can't necessarily assume the same for provider owned keys. For now though we are explicitly downgrading EC keys prior to point encoding/decoding in order to use the ameth implementation. It seems we could do the same trick here and then just do it the old way - until such time as the whole thing goes provider side.

File: doc/man3/EVP_PKEY_is_a.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_is_a.pod]: I assume this returns false if the key can only be used for verifying? e.g. because it is a public key? Could we make that more explicit?

File: doc/man3/EVP_PKEY_set1_RSA.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_set1_RSA.pod]: Should we attempt a downgrade if someone calls these functions?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: A pointer is just a value just like any other value.  That value won't "magically" change because you freed the memory area its value points at.  The value in itself is perfectly usable as long as it's not dereferenced.

But, I can waste another variable to hold the truth value if you prefer...

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: Looking at the legacy code, it seems we assume that method that supports signing also supports verifying.  Looking at what `EV_KEY_can_sign()` does, I found these indicators:

https://github.com/openssl/openssl/blob/20c98cd45399423f760dbd75d8912769c6b7b10e/crypto/ec/ec_local.h#L31-L32

https://github.com/openssl/openssl/blob/20c98cd45399423f760dbd75d8912769c6b7b10e/include/openssl/evp.h#L1402-L1404

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: Okie

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: I wasn't ready to assume, so...

(I assume, BTW, that we're going to look at *all* TODO comments before final release, and that the presence of "(3.0)" only indicates if it's a "MUST do" or a "MAY do")

File: doc/man3/EVP_PKEY_is_a.pod
Comment (levitte) [reply doc/man3/EVP_PKEY_is_a.pod]: Ah, so having it verify that `evp_keymgmt_has(p->keymgmt, p->keydata, OSSL_KEYMGMT_SELECT_PRIVATE_KEY)` is true?  Yeahsure

File: doc/man3/EVP_PKEY_set1_RSA.pod
Comment (levitte) [reply doc/man3/EVP_PKEY_set1_RSA.pod]: I'm unsure...  but as a side thought, I've been thinking that keys with a provider side "origin" should have the type `EVP_PKEY_KEYMGMT` rather than `EVP_PKEY_NONE`...

File: crypto/evp/p_lib.c
Comment (richsalz) [reply crypto/evp/p_lib.c]: FWIW, I agree that looking at the pointer after it's free'd, while technically legal, is confusing.

Comment (kaduk) [reply crypto/evp/p_lib.c]: I think that *technically* it's not actually legal.
It helps to recall that the C rules for pointers have their origins in the era of segmented architectures, where even trying to load a pointer value (not dereference it) for which the corresponding segment was not valid anymore, could fault.

Comment (levitte) [reply crypto/evp/p_lib.c]: Huh!  I didn't know that.

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: Should we care in the legacy section of this function?  How do we check if there's a private key there?

File: crypto/evp/p_lib.c
Comment (t8m) [reply crypto/evp/p_lib.c]: You could nicely do 
```
if (signature != NULL) {
    EVP_SIGNATURE_free(signature); 
    return 1; 
} return 0;
```

Comment (richsalz) [reply crypto/evp/p_lib.c]: I believe Ben is wrong; "free(p); look at p but not *p" is legal.

Comment (kaduk) [reply crypto/evp/p_lib.c]: See n1256.pdf appendix J.2, stanzas 8 and 9:

```
— An object is referred to outside of its lifetime (6.2.4).
— The value of a pointer to an object whose lifetime has ended is used (6.2.4).
```
The first one refers to dereferencing the pointer, and the second refers to only the value of the (undereferenced) pointer.

Comment (richsalz) [reply crypto/evp/p_lib.c]: I stand corrected, thanks.

File: ssl/t1_lib.c
Comment (levitte) [reply ssl/t1_lib.c]: Okie, done

File: doc/man3/EVP_PKEY_set1_RSA.pod
Comment (levitte) [reply doc/man3/EVP_PKEY_set1_RSA.pod]: I'm leaning on no.  We risk too much uncontrolled downgrading.

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: The question here remains unanswered.  @mattcaswell

Comment (mattcaswell) [reply crypto/evp/p_lib.c]: It seems all this stems from confusion in my own head about what EVP_PKEY_can_sign() is supposed to do. Is it supposed to report whether the parameters associated with this key are capable of doing signing operations, or whether this specific key can do signing (i.e. it has a private key *and* it has suitable parameters). I was assuming the latter - but if this is supposed to be a drop-in replacement for EC_KEY_can_sign() then that seems to do the former - and therefore don't care whether this has a private key or not.

File: doc/man3/EVP_PKEY_is_a.pod
Comment (mattcaswell) [reply doc/man3/EVP_PKEY_is_a.pod]: See my comment above...

File: ssl/t1_lib.c
Comment (mattcaswell) [inline ssl/t1_lib.c]: Now that we have downgraded we can keep all the original code below. The `EVP_PKEY_get0_EC_KEY()` call will just work and we don't need `evp_pkey_get_EC_KEY_conv_form` or `evp_pkey_get_EC_KEY_field_type`.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c]: I've designed it as a drop in replacement for `EC_KEY_can_sign()`.  It's possible that we want to make it stricter before releasing 3.0, though...  but maybe not immediately, as we seem to get ourselves into another philosophical rabbit hole.

File: ssl/t1_lib.c
Comment (levitte) [reply ssl/t1_lib.c]: At *some* point, we will need to get those things from provider side keys, so I'm apprehensive at removing those functions...  do you anticipate that the *final solution* will look substantially different?

File: crypto/evp/p_lib.c
Comment (levitte) [reply crypto/evp/p_lib.c:777]: Fixed

File: ssl/t1_lib.c
Comment (levitte) [reply ssl/t1_lib.c:593]: Ping @mattcaswell

Comment (mattcaswell) [reply ssl/t1_lib.c:593]: Yes. This whole dance with these calls is just to figure out what TLS compression id we need to use. I anticipate asking the provider directly for this compression id rather than inferring it from these other calls.

Comment (kaduk) [reply ssl/t1_lib.c:593]: (I assume "TLS compression id" means "EC point format", not "we're using something other than NULL compression".)

Comment (mattcaswell) [reply ssl/t1_lib.c:593]: Yes...the TLS EC point format. Sorry for the confusion.

Comment (levitte) [reply ssl/t1_lib.c:593]: In that case, I might as well restore this whole function to its original form, and just replace the EVP_PKEY_id() check with a EVP_PKEY_is_a() check.

Comment (levitte) [reply ssl/t1_lib.c:593]: Done

**Comment by levitte** [comment]
Ah, the CI failure was due to a merge error with duplicate definition of the same local variable.
Fixed, and I took the opportunity to squash and rebase.

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Approved assuming CIs pass.

**Comment by levitte** [comment]
... it only takes a rebase for hell to break loose :unamused:

**Comment by levitte** [comment]
With @mattcaswell's assistance, I was able to put in a workable hack.  It turned out that legacy key support needed some hard coding.

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
One failure left...  I didn't account for `no-ec`

**Comment by levitte** [comment]
... at least not everywhere

**Comment by levitte** [comment]
Fixed

**Review by mattcaswell — APPROVED** [review]
Reconfirm assuming CIs are ok

**Comment by levitte** [comment]
Apart from the usual timeout, CIs are fine

**Comment by levitte** [comment]
Merged.

4f76d62f2384d3335bd1d043706995ae64b37348 EVP: add EVP_PKEY_is_a() and EVP_PKEY_can_sign()
c2041da8c15027ddde5afcf9809d8d3a975eb25b EVP & TLS: Add necessary EC_KEY data extraction functions, and use them
e3be0f4389fe1680e1fd547a8575ea71b4518b57 Fix export of provided EC keys
afce590b74159f7df1452fb2c4aa990a52536c38 TLS: Temporarly downgrade newly generated EVP_PKEYs to legacy

**Comment by levitte** [comment]
Oh damn, *now* I see that the last approval came after the "approval: done" label was set, so I cut the grace period short...  by 3hrs, I hope it's not too bad.

**Pull request closed without merging**
